femwell/pn_analytical.py

```diff
@@ -0,0 +1,199 @@
+"""
+
+import pickle
+
+import numpy as np
+
+
+def dn_carriers(wavelength: float, dN: float, dP: float) -> float:
+    """Phenomenological wavelength-dependent index perturbation from free carriers in silicon.
+
+    Use quadratic fits for wavelengths, or better-characterized models at 1550 nm and 1310 nm.
+
+    Args:
+        wavelength: (um).
+        dN: excess electrons (/cm^3).
+        dP: excess holes (/cm^3).
+
+    Returns:
+        dn: change in refractive index compared to intrinsic silicon.
+    """
+    if wavelength == 1.55:
+        return -5.4 * 1e-22 * np.power(dN, 1.011) - 1.53 * 1e-18 * np.power(dP, 0.838)
+    elif wavelength == 1.31:
+        return -2.98 * 1e-22 * np.power(dN, 1.016) - 1.25 * 1e-18 * np.power(dP, 0.835)
+    else:
+        wavelength *= 1e-6
+        return -3.64 * 1e-10 * wavelength**2 * dN - 3.51 * 1e-6 * wavelength**2 * np.power(
+            dP, 0.8
+        )
+
+
+def dalpha_carriers(wavelength: float, dN: float, dP: float) -> float:
+    """Phenomenological wavelength-dependent absorption perturbation from free carriers in silicon.
+
+    Use quadratic fits for wavelengths, or better-characterized models at 1550 nm and 1310 nm.
+
+    Args:
+        wavelength: (um).
+        dN: excess electrons (/cm^3).
+        dP: excess holes (/cm^3).
+
+    Returns:
+        dalpha: change in absorption coefficient (/cm)
+    """
+    if wavelength == 1.55:
+        return 8.88 * 1e-21 * dN**1.167 + 5.84 * 1e-20 * dP**1.109
+    elif wavelength == 1.31:
+        return 3.48 * 1e-22 * dN**1.229 + 1.02 * 1e-19 * dP**1.089
+    else:
+        wavelength *= 1e-6
+        return 3.52 * 1e-6 * wavelength**2 * dN + 2.4 * 1e-6 * wavelength**2 * dP
+
+
+def alpha_to_k(alpha, wavelength):
+    """Converts absorption coefficient (/cm) to extinction coefficient (unitless), given wavelength (um)."""
+    wavelength = wavelength * 1e-6  # convert to m
+    alpha = alpha * 1e2  # convert to /m
+    return alpha * wavelength / (4 * np.pi)
+
+
+def k_to_alpha(k, wavelength):
+    """Converts extinction coefficient (unitless) to absorption coefficient (/cm), given wavelength (um)."""
+    wavelength = wavelength * 1e-6  # convert to m
+    alpha = 4 * np.pi * k / wavelength
+    return alpha * 1e-2  # convert to /cm
+
+
+def k_to_alpha_dB(k, wavelength):
+    """Converts extinction coefficient (unitless) to absorption coefficient (dB/cm), given wavelength (um)."""
+    wavelength = wavelength * 1e-6  # convert to m
+    alpha = 4 * np.pi * k / wavelength
+    return 10 * np.log10(np.exp(1)) * alpha * 1e-2  # convert to /cm
+
+
+# Physical constants (in cm)
+q = 1.60217663e-19  # elementary charge (C)
+eps = 11.68 * 8.854e-12 / 100  # relative permittivity of silicon (CV−1m−1 * 1m / 100 cm)
+kB = 1.380649e-23  # Boltzmann constant (J/K)
```

Comment:
could we use for q and kb directly scipy.constans? (just `from scipy.constans import ...`)

Reply:
yes but we need to convert the lengths to cm

Reply:
Okay, then just add the factor - this conversion should I think always be explicitly stated.
I'm also planning on updating the mode-solver, that it becomes scale invariant. We can then give it a try and see if we can do everything in SI-units (to avoid such conversions and possible errors) or if it decreases the precision